### PR TITLE
docs(rfc): refine the regression testing suite FE-2637

### PR DESCRIPTION
**Pull request by sianford**
### Description

This draft RFC details proposed changes to how we regression test the Carbon components in Storybook, using mainly visual comparison testing, and which will require a reformatting of all the stories. This will facilitate running regression at every commit.

### CODE REVIEW

**Comment by DlgSHi** [comment]
Is it still a draft as in description?

**Comment by sianford** [comment]
It’s a draft in the sense that it’s the first version of this and is open for debate and amendment.

File: rfcs/text/testing-refinement.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+By presenting the various different pre-set configurations of the components in as few frames and stories as possible without the need to set knobs, the tests could run faster while Applitools could take and compare a vastly reduced number of snapshots. The purpose of our Carbon component library is to showcase to Developers how they can create a beautiful UI in a fast and easy manner with our components, not how to amend knobs in Storybook.
+
+# Detailed design
+
+To facilitate a change to this method of regression testing, the components need to be presented differently in Storybook with all possible configurations pre-set in the stories and no knobs to amend. This has already been done to some extent for some components e.g. Button in the `Design System` directory. If all of the configurations and scenarios can be demonstrated in stories in the `Design System` directory then they should, with regression testing run against those stories. Any additional cases, such as edge cases or extended test cases e.g. tests for special characters in labels, should be added in a story in the `Test` directory therefore using the `Test` directory purely for cases for extended testing only.
+
```
Comment (ianforster-sage) [inline rfcs/text/testing-refinement.md:41]: Would a screenshot be useful here? The Button example where you can see when using a certain icon the left alignment is out would illustrate it well.

Comment (sianford) [reply rfcs/text/testing-refinement.md:41]: Yes, I'll add that. At the point I created this, the All Buttons story showing all the button examples including all icons hadn't been created.

File: rfcs/text/testing-refinement.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+- [How we teach this](#how-we-teach-this)
+- [Unresolved questions](#unresolved-questions)
+
+# Summary
+
+The Carbon team would like to use a more rapid and accurate method of regression testing Carbon components to reduce development to release time and enable the regression suite to be run at every commit.
+
+# Motivation
+
+The regression suite uses the Cypress.io framework to test all knob configurations, events and accessibility of all stories of all components in Storybook. In its current form it takes around 3.5 hours to run. When the build itself only takes around 7 minutes, it is disproportionate to have a regression suite for those same components to take that much longer to run by such a huge factor. We require a method by which the visual UI can be tested rapidly, accurately and at every commit to detect regressions earlier.
```
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:26]: Yes, I'll change that and the first instance of "Applitools" to a link.

File: rfcs/text/testing-refinement.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+## Visual Comparison testing
+Making assertions of values in the CSS is not the most reliable way to verify the visual styling of the components is correct. In addition, this has been found to be brittle in cases such as where browser upgrades slightly amend the number of pixels of component dimensions and the test spec is asserting on specific pixel values.
+
+The visual comparison testing tool ‘Applitools’ has been under test for some time. It is probably the most mature of this type of tool and is endorsed by prominent members of the testing community. Snapshots of the required screen or frame are taken and compared with baseline snapshots. Any differences that can be detected with the human eye are highlighted as test failures. While we are only currently primarily concerned with testing the components in Chrome, we do have the option with Applitools to render the snapshots of the DOM in almost any other operating system, browser, device and viewport configuration we like by using their Ultrafast Grid. This would give us an early indication of any styling issues with the components in other set ups.
```
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:31]: Yep, I'll change that to a link and remove the quotes.

File: rfcs/text/testing-refinement.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+Recently, we have carried out some proof of concept testing of the Carbon components by integrating Applitools with our existing Cypress regression suite. While we can see that this would be a more accurate way of testing the visual attributes of the components, this activity has highlighted that in its current form, the regression suite with Applitools integration is not a viable option as the vast number of comparisons extend an already lengthy suite.
+
```
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:34]: I'll add some details about the number of feature files that were amended to add steps to add Applitools snapshot captures and how long it took to run them.

File: rfcs/text/testing-refinement.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  - [Reformatting of component stories for testing](#reformatting-of-component-stories-for-testing)
+- [Detailed design](#detailed-design)
+  - [Test stories](#test-stories)
+  - [Test feature files](#test-feature-files)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Adoption strategy](#adoption-strategy)
+- [How we teach this](#how-we-teach-this)
+- [Unresolved questions](#unresolved-questions)
+
```
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:19]: Considering the wider audience that might view this document, it might be useful. I'll add one.

File: rfcs/text/testing-refinement.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    When I click on "button"
+    Then click action was called in Actions Tab
+```    
+
+The step “Then Button is presented correctly” is where Applitools will take the snapshot. The existing accessibility and build test feature files could remain as they are.
+
+During initial proof of concept testing of Applitools in 2019, some basic tests were executed using Cypress to navigate to some component URLs and for Applitools to take a snapshot. The typical total time for Applitools to snapshot one component, render the DOM in 10 different configurations and return the results was around 50 seconds, so if we just wanted to snapshot 1 URL and compare in 1 config this should take around 5 seconds per story. There are currently around 150 stories (plus 18 for deprecated components although these will disappear at some point). This would take an estimated 12.5 minutes to run in the proposed set up. For some components, more stories may be required and also more components will be created adding to the numbers but for many the number of stories could be reduced to 1. The estimated time to capture and compare snapshots will be around these 12.5 minutes of estimated time.
```
Comment (jamime) [inline rfcs/text/testing-refinement.md:74]: > There are currently around 150 stories (plus 18 for deprecated components although these will disappear at some point)

I think this will actually be far fewer. We have multiple stories for each component at the moment, but we intend to comine them into one snapshot per component.

File: rfcs/text/testing-refinement.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+# Adoption strategy
+
+Developers would need to amend how they write the stories. As many examples as possible of various main configurations of the components will need to be included in one story without any knobs. Other React libraries such as Material-UI (https://material-ui.com) use this method of demonstrating their components. The current standard of documentation in the `Docs` tab will need to be maintained to provide a thorough description of how the component is meant to be used.
+
+QA Engineers will amend existing specs to leave just the steps to capture snapshots with Applitools. QA Engineers will need to identify additional scenarios and use cases not included in the main `Design System` directory as currently. These would be added to the `Test` directory, although it would be beneficial if QA Engineers were able to add these themselves with the help of Developers if required.
```
Comment (jamime) [inline rfcs/text/testing-refinement.md:100]: It would be best to prefer unit tests, adding stories only when we want to do a visual comparison.

File: rfcs/text/testing-refinement.md (hunk 6 above)
Comment (damianOfKyan) [reply rfcs/text/testing-refinement.md:74]: one snapshot to rule them all.

Comment (sianford) [reply rfcs/text/testing-refinement.md:74]: Yes, that's what I'm trying to get over here and also in the sub-section 'Reformatting of component stories for testing'. A reduction of stories with less snapshots.

**Comment by sianford** [comment]
@jamime I've received information from Applitools Support. It's probably not what you were hoping for though. In answer to your questions:

1. The user limit for our account is 3 and we're currently maxed out. We may be able to get more capacity by conversing with whoever from Applitools set the account up. I'll try to find out who that is.
2. Anyone with access to the PR can see the test results. However, they will not be able to access the URL to the actual test results on the Applitools dashboard. They will only the see the pass or fail on the PR.

File: rfcs/text/testing-refinement.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+The test feature file for Button could be reduced to something like:
+
+```
+Feature: Button component
+  I want to check Button component properties
+
+  Background: Open Button component default page
+    Given I open "Button" component page
+
```
Comment (DlgSHi) [inline rfcs/text/testing-refinement.md:79]: we need to mention about tags here as you mentioned below:
```suggestion
@ applitools
```

File: rfcs/text/testing-refinement.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+One advantage of integrating Applitools is having the option of rendering the DOM snapshots in many other operating system, browser, device and viewport configurations we like by using their Ultrafast Grid. This does not mean though that we are testing our components in these other configurations at the time of test execution. There are other vendors, including CrossBrowserTesting and BrowserStack, which would allow us to run our required tests directly in these other configurations of real machines. However, this would mean a change of automation framework from Cypress to Selenium. This seems like a drastic change at this point as all the files, including the support files like helper files, indexes, locators etc, would need rewriting. In addition, not all of these other products are free for open source repositories as Applitools is and don’t all provide a dashboard which is as user-friendly as that of Applitools.
```
Comment (DlgSHi) [inline rfcs/text/testing-refinement.md:117]: I don't think we will change our framework for cypress to selenium

File: rfcs/text/testing-refinement.md (hunk 7 above)
Comment (sianford) [reply rfcs/text/testing-refinement.md]: I've added something to that effect in the newer version I've pushed for review again.

File: rfcs/text/testing-refinement.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+- [How we teach this](#how-we-teach-this)
+- [Unresolved questions](#unresolved-questions)
+
+
+# Glossary
+
+| Term           | Description                                                                    |
+|----------------|--------------------------------------------------------------------------------|
+| Applitools     | An automated visual UI testing tool                                            |
+| canvas         | An area in the preview panel in Storybook                                      |
```
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:27]: I only capitalised proper nouns. Is that not correct?

File: rfcs/text/testing-refinement.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+| CSS            | Cascading Style Sheets, to describe how HTML elements are to be displayed      |
+| Cypress        | An open source frontend testing framework                                      |
+| Design System  | Guidelines around use, styles, behaviours etc for developing Carbon components |
+| knobs          | A means of editing component props dynamically in the Storybook UI             |
```
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:31]: As above, I only capitalised proper nouns.

File: rfcs/text/testing-refinement.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+| props          | Properties of the component which affect how it appears/behaves when rendered  |
```
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:32]: As above, I only capitalised proper nouns. Is that not correct?

File: rfcs/text/testing-refinement.md [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+Recently, we have carried out some proof of concept testing of the Carbon components by integrating Applitools with 18 component test files from our existing Cypress regression suite. Steps were added into the feature files and relevant support files to enable Applitools to take a snapshot after each test scenario. While we can see that this would be a more accurate way of testing the visual attributes of the components, this activity has highlighted that in its current form, the regression suite with Applitools integration is not a viable option as the vast number of comparisons extend an already lengthy suite.
```
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:48]: There were 18 feature files tested, hence why I used the word "files" and not "cases".

File: rfcs/text/testing-refinement.md [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+Recently, we have carried out some proof of concept testing of the Carbon components by integrating Applitools with 18 component test files from our existing Cypress regression suite. Steps were added into the feature files and relevant support files to enable Applitools to take a snapshot after each test scenario. While we can see that this would be a more accurate way of testing the visual attributes of the components, this activity has highlighted that in its current form, the regression suite with Applitools integration is not a viable option as the vast number of comparisons extend an already lengthy suite. 
+
+## Reformatting of component stories for testing
+By presenting the various different pre-set configurations of the components in as few preview frames and stories as possible without the need to set knobs, the tests could run faster while Applitools could take and compare a vastly reduced number of snapshots. The purpose of our Carbon component library is to showcase to Developers how they can create a beautiful UI in a fast and easy manner with our components, not how to amend knobs in Storybook.
```
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:51]: I added this complete sentence to explain to readers that changing knobs in Storybook is not testing the components. It's testing that Storybook knobs work. Therefore, we can showcase our components and test that they visually are correct by using Applitools for visual comparison testing. There won't be any point to the sentence with that part removed.

File: rfcs/text/testing-refinement.md [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+
+# Detailed design
+
+To facilitate a change to this method of regression testing, the components will need to be presented differently in Storybook in stories specific for this purpose with all possible configurations pre-set in the stories with few to no knobs to amend. This has already been done to some extent for some components e.g. Button in the `Design System` directory.
```
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:55]: Isn't this only for stories in the Design System directory though? The ones for visual comparison testing with all examples, where possible, will be in the Test directory like the 'All Buttons' story (once it is moved) which isn't an .mdx file. It's just in the button.stories.js file.

File: rfcs/text/testing-refinement.md (hunk 9 above)
Comment (sianford) [reply rfcs/text/testing-refinement.md:117]: Agreed, at least for the foreseeable future. I had to list it as an alternative though.

File: rfcs/text/testing-refinement.md [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+| props          | Properties of the component which affect how it appears/behaves when rendered  |
+| Storybook      | An open source frontend tool for developing and organising UI components       |
+
+# Summary
+
+The Carbon team would like to use a more rapid and accurate method of regression testing Carbon components to reduce development to release time and enable the regression suite to be run at every commit.
+
+# Motivation
+
+The regression suite uses the [Cypress.io](https://www.cypress.io) framework to test all knob configurations, events and accessibility of all stories of all components in Storybook. In its current form it takes around 2.5 hours to run. When the build itself only takes around 7 minutes, it is disproportionate to have a regression suite for those same components to take that much longer to run by such a huge factor. We require a method by which the visual UI can be tested rapidly, accurately and at every commit to detect regressions earlier.
```
Comment (DlgSHi) [inline rfcs/text/testing-refinement.md:41]: according to [this change](https://github.com/Sage/carbon/pull/2880) regression time will be under 30 min

File: rfcs/text/testing-refinement.md [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+
+![CARBON](./../../.assets/button_design_system.png)
+
+and the `All Buttons` story for testing purposes
+
+![CARBON](./../../.assets/all_buttons_story.png)
+
+Existing functional or event tests which Cypress is required for will need to be amended to test in the `Docs` tab of the new refactored stories rather than in the `Canvas`. The accessibility tests should not be affected. Any visual verification tests currently done by Cypress alone should be done by Applitools. Any additional cases, such as edge cases, cases that cannot be tested in the component in the `Design System` directory or extended test cases e.g. tests for special characters in labels, should be added in an additional story in the `Test` directory. If all of the configurations can be demonstrated in a single story in the `Test` directory then they should, with visual regression testing run against those stories. We should, however, take care to test only what we deem as bringing value to our regression testing and not have Applitools snapshot everything just because we can.
```
Comment (DlgSHi) [inline rfcs/text/testing-refinement.md:63]: e.g. tests for special characters in labels - here should be also mentioned other tests we should perform such as: keyboard accessibility, action events tests

File: rfcs/text/testing-refinement.md [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+
+![CARBON](./../../.assets/material-ui-argos-buttons-example.png)
+
+In this example alone, 13 test cases in our current regression suite could be reduced to just 1 snapshot from Applitools.
+
+## Test feature files
+Cypress will use the feature files to navigate to each story, as it does currently, but without setting all the configuration via the knobs. A step will be added for Applitools to snapshot the canvas. This will remove the need for the numerous steps taken just to get the component into the pre-requisite state on which to assert. The regression suite could be reduced to one Cypress test file for each component. Any differences between this and the baseline snapshot would comprise a test failure.
```
Comment (DlgSHi) [inline rfcs/text/testing-refinement.md:73]: The regression suite could be reduced to one Cypress test file for each component. - it is already implemented [here](https://github.com/Sage/carbon/pull/2880)

File: rfcs/text/testing-refinement.md [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+
+# Alternatives
+
+There is some duplication between the Cypress regression tests and the unit tests. Indeed, the act of setting knobs to various states to set a prerequisite state and the result asserted on is already done in many of the unit tests. An exercise to remove duplicate tests in the regression suite could be carried out to speed up the run time. The recent removal of support of the ‘classic’ theme and related feature files reduced the regression suite run time by 1-1.5 hours. We could upgrade our Cypress package to a paid for service or move to the open source OSS plan which would allow the tests to be parallelised rather than them all running in one thread, reducing the run time further. All of these actions though would not reduce the regression suite run time by the required factor nor provide the visual comparison testing capability that Applitools integration would.
```
Comment (DlgSHi) [inline rfcs/text/testing-refinement.md:123]: Upgrading cypress to OSS plan is done [here](https://github.com/Sage/carbon/pull/2880) and we already mentioned about this before, so no need to mention about it in this section

File: rfcs/text/testing-refinement.md [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+
+Applitools can integrate with Jira and automatically log a defect when a test has failed which could be useful and save time with manual intervention, although initial investigation into whether the failure is a valid one should probably be done first so this may not be a feature we could adopt immediately.
+
+# How we teach this
+
+The Testing Strategy `README.md` will be updated with details of this new process. As creating the automation tests now form part of the definition of done for any new components, Developers may need some assistance with understanding how to write the new format of Cypress tests. Support documentation for Applitools is available from https://applitools.com to help with Cypress integration and GitHub. Assistance can also be provided by the Automation Team who have been undertaking the most recent phase of Applitools integration testing with Cypress.
```
Comment (DlgSHi) [inline rfcs/text/testing-refinement.md:137]: we should provide correct path to`README.md` file in cypress directory for testing strategy

File: rfcs/text/testing-refinement.md (hunk 17 above)
Comment (jamime) [inline rfcs/text/testing-refinement.md:63]: > Existing functional or event tests which Cypress is required for will need to be amended to test in the Docs tab of the new refactored stories rather than in the Canvas.

This is not correct. You can still use the Canvas tab if the story has been written in MDX. There are issues if reference anoter story using a story-id. But we'll be using MDX going forwards.

File: rfcs/text/testing-refinement.md [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
...ations of the components should be included in one story in the `Design System` directory without any knobs where possible. Existing stories will need to be amended to this format. Other React libraries such as Material-UI (https://material-ui.com) use this method of demonstrating their components. The current standard of documentation in the `Docs` tab will need to be maintained to provide a thorough description of how the component is meant to be used. In addition, these stories will need to be amended where necessary to show events in the `Actions` tab. Developers will need to write one or more additional stories in the `Test` directory for testing purposes presenting as many examples as possible of the components for visual regression purposes. The Cypress tests will need to be compared with the unit tests and, where tests covered by the regression suite can be moved into the unit tests, this should be done. Whilst some duplication will remain where event tests are concerned, we feel it is necessary to have these tests in Cypress too in order to provide a second layer of testing and further confidence that the action triggered by a user is the correct one, so these will remain.
```
Comment (jamime) [inline rfcs/text/testing-refinement.md:129]: > Developers will need to write one or more additional stories in the Test directory for testing purposes presenting as many examples as possible of the components for visual regression purposes. 

I think we said that we were going to snapshot the MDX/Docs page which shows all examples of each component? The only reason to add a `Test` story is to do a visual regression or additional cypress (unit) test that we do not want to be in our Documentation.

File: rfcs/text/testing-refinement.md (hunk 16 above)
Comment (sianford) [reply rfcs/text/testing-refinement.md:41]: The problem is that these two pieces of work have been done in parallel, so any changes I make in this RFC are out of date when changes are done in the other stream. I'm going to change this section altogether as the motivation for integrating Applitools now is no longer to reduce the time taken, only to provide the visual UI testing.

File: rfcs/text/testing-refinement.md (hunk 17 above)
Comment (sianford) [reply rfcs/text/testing-refinement.md:63]: Didn't we agree that action events tests could be done from the stories in `Design System` where possible as those stories would be changed to match what was done in one component recently to allow that? Also, I thought the accessibility tests would continue to be done where they currently are now, with most of the components being tested from `Design System` or the main component list? I thought we'd agreed that only those tests that could not be done from `Design System` and additional specific visual testing would be done in the `Test` directory?

File: rfcs/text/testing-refinement.md (hunk 18 above)
Comment (sianford) [reply rfcs/text/testing-refinement.md:73]: But we have multiple feature files for some components including `Button` and `Icon`? `Icon` in particular could be one single file called 'All Icons' like the 'All Buttons' one we have.

File: rfcs/text/testing-refinement.md (hunk 19 above)
Comment (sianford) [reply rfcs/text/testing-refinement.md:123]: As commented above, these two streams of work have been done in parallel and that work to implement the OSS plan was not before this RFC was created/updated.

File: rfcs/text/testing-refinement.md (hunk 21 above)
Comment (sianford) [reply rfcs/text/testing-refinement.md:129]: The current MDX/Docs pages don't show all examples of each component though. Like `Button`, hence why the 'All Buttons' story was created for testing purposes. Not all components have been updated to the format of the others in the Design System directory yet.`Icon` for example has only one example until you change the knobs and we can't do one snapshot for every single icon. Will every icon be shown once it's been refactored in the `Design System` directory? If not, then another story, like the 'All Buttons' story, will need to be written to show all icons where one single snapshot can be take. That's what I'm saying here.

**Comment by sianford** [comment]
> @jamime I've received information from Applitools Support. It's probably not what you were hoping for though. In answer to your questions:
> 
> 1. The user limit for our account is 3 and we're currently maxed out. We may be able to get more capacity by conversing with whoever from Applitools set the account up. I'll try to find out who that is.
> 2. Anyone with access to the PR can see the test results. However, they will not be able to access the URL to the actual test results on the Applitools dashboard. They will only the see the pass or fail on the PR.

@jamime Applitools have since opened up the max user limit to allow all of the team access. They've also made the account public so third parties can view the results too.

File: rfcs/text/testing-refinement.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:27]: I've added further entries to the Glossary and amended the description of Canvas and Docs to be names of tabs, these are now proper nouns so I've capitalised them. As only 'knobs' and 'props' remained lower case, I've capitalised them anyway as they now look odd in a longer list of capitalised entries.

File: rfcs/text/testing-refinement.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:31]: I've added further entries to the Glossary and amended the description of Canvas and Docs to be names of tabs, these are now proper nouns so I've capitalised them. As only 'knobs' and 'props' remained lower case, I've capitalised them anyway as they now look odd in a longer list of capitalised entries.

File: rfcs/text/testing-refinement.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md:32]: I've added further entries to the Glossary and amended the description of Canvas and Docs to be names of tabs, these are now proper nouns so I've capitalised them. As only 'knobs' and 'props' remained lower case, I've capitalised them anyway as they now look odd in a longer list of capitalised entries.

File: rfcs/text/testing-refinement.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md]: I've since rewritten that section.

File: rfcs/text/testing-refinement.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md]: I've since rewritten this section.

File: rfcs/text/testing-refinement.md [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+
+![CARBON](./../../.assets/button_design_system.png)
+
+These changes will not impact stories in the `Design System` directory, which are there to showcase the components and provide information about their use and behaviour, nor will it affect any other functional, event or accessibility tests which Cypress alone is required for. Any visual verification tests currently done by Cypress should be done by Applitools instead. Any additional cases, such as edge cases, cases that cannot be tested in the component in the `Design System` directory or extended test cases e.g. tests for special characters in labels, should be added in an additional story in the `Test` directory. If all of the configurations can be demonstrated in a single story in the `Test` directory then they should, with visual regression testing run against those stories. We should, however, take care to test only what we deem as bringing value to our regression testing and not have Applitools snapshot everything just because we can.
```
[the remark this answers is not in the document]
Comment (sianford) [reply rfcs/text/testing-refinement.md]: I've since rewritten this section.

**Comment by sianford** [comment]
Will resume this work once parallel changes to the Cypress framework and CI changes have been implemented, upon which time a new pull request will be created.
